xarray/core/coordinates.py

```diff
@@ -796,6 +789,114 @@ def _ipython_key_completions_(self):
         ]
 
 
+class DataTreeCoordinates(Coordinates):
+    """
+    Dictionary like container for coordinates of a DataTree node (variables + indexes).
+
+    This collection can be passed directly to the :py:class:`~xarray.Dataset`
+    and :py:class:`~xarray.DataArray` constructors via their `coords` argument.
+    This will add both the coordinates variables and their index.
+    """
+
+    # TODO: This only needs to be a separate class from `DatasetCoordinates` because DataTree nodes store their variables differently
+    # internally than how Datasets do, see https://github.com/pydata/xarray/issues/9203.
+
+    # TODO should inherited coordinates be here? It would be very hard to allow updating them...
+    # But actually maybe the ChainMap approach would make this work okay??
+
+    _data: DataTree
```

Comment:
Inheriting from `Coordinates` and using `DataTree` instead of `DataWithCoords` here both expose a few legitimate typing issues. Particularly:
- `DataTree` doesn't directly inherit from `DataWithCoords`, though maybe it could inherit from a future version of it (see https://github.com/pydata/xarray/issues/9203#issuecomment-2218309083)
- `Hashable` vs `str` annoyances (#8836)

Mostly we have got away with these discrepancies but inheritance here is probably going to mean I need to either `ignore` or `cast` all over the place?

cc @headtr1ck

Reply:
actually this wasn't very difficult to work around, it mostly just required adding a `type: ignore[assignment]` here :)